### PR TITLE
If possible use symbol name instead of entry name for function name

### LINKED ISSUE #15260: Some functions not labeled properly in Android shared libraries
**Environment information**
* Operating System: Kali 
* Cutter version: 1.8.3 
* File format: Shared Library (.so) 

**Describe the bug**

I've been using Cutter to solve some Crackmes, I ran into this issue with [this one](https://github.com/OWASP/owasp-mstg/blob/master/Crackmes/Android/Level_03/UnCrackable-Level3.apk) specifically. If you unpack or unzip the APK you will find a library in the lib folder called libfoo.so, which has some dynamically linked Java methods. Java methods in native libraries called by Android apps need to be named as such `Java_com_android_interesting_Stuff_doThingsInNativeLibrary`. In libfoo.so, Cutter detects two of the native methods properly but does not detect `Java_sg_vantagepoint_uncrackable3_MainActivity_init`. If I search for the function I can find it, but on the function bar on the left, it is named entry.init0.

**Expected behavior**

Other disassemblers (like Ghidra) preserve the original name of the function when listing functions, I think that Cutter should follow this behavior as well.

### PR DESCRIPTION
**Pull request by pelijah**
**Your checklist for this pull request**
- [ ] I've read the [guidelines for contributing](https://github.com/radareorg/radare2/blob/master/DEVELOPERS.md) to this repository
- [ ] I made sure to follow the project's [coding style](https://github.com/radareorg/radare2/blob/master/DEVELOPERS.md#code-style)
- [ ] I've added tests that prove my fix is effective or that my feature works (if possible)
- [ ] I've updated the documentation and the [radare2 book](https://github.com/radareorg/radare2book) with the relevant information (if needed)

**Detailed description**

...

**Test plan**

...

**Closing issues**

Closes #15260

### CODE REVIEW

File: libr/core/canal.c [hunk 1]
```diff
@@ -819,8 +823,20 @@ static int __core_anal_fcn(RCore *core, ut64 at, ut64 from, int reftype, int dep
 		} else if (fcnlen == R_ANAL_RET_END) { /* Function analysis complete */
 			f = r_core_flag_get_by_spaces (core->flags, fcn->addr);
 			if (f && f->name && strncmp (f->name, "sect", 4)) { /* Check if it's already flagged */
-				R_FREE (fcn->name);
-				fcn->name = strdup (f->name);
+				const char *new_name = f->name;
```
Comment (radare) [inline libr/core/canal.c:826]: what about using the rbin api instead of the flags one?

Comment (pelijah) [reply libr/core/canal.c:826]: Do you mean `r_bin_get_symbols()`?

File: libr/core/canal.c [hunk 2]
```diff
@@ -819,8 +823,20 @@ static int __core_anal_fcn(RCore *core, ut64 at, ut64 from, int reftype, int dep
...
+				if (is_entry_flag (f)) {
+					RListIter *iter;
+					RFlagItem *flag;
+					const RList* flags = r_flag_get_list (core->flags, fcn->addr);
+					r_list_foreach (flags, iter, flag) {
+						if (r_str_startswith (flag->name, "sym.")) {
```
Comment (thestr4ng3r) [inline libr/core/canal.c:832]: Don't check the prefix, you can do something like
```c
if (flag->space && !strcmp (flag->space->name, R_FLAGS_FS_SYMBOLS))
```

File: libr/core/canal.c [hunk 3]
```diff
@@ -741,6 +741,10 @@ static void set_fcn_name_from_flag(RAnalFunction *fcn, RFlagItem *f, const char
 	}
 }
 
+static bool is_entry_flag(RFlagItem *f) {
+	return f->space && !strcmp (f->space->name, "symbols") && r_str_startswith (f->name, "entry.");
```
Comment (thestr4ng3r) [inline libr/core/canal.c:745]:
```suggestion
	return f->space && !strcmp (f->space->name, R_FLAGS_FS_SYMBOLS) && r_str_startswith (f->name, "entry.");
```

File: libr/core/canal.c (hunk 1 above)
Comment (radare) [reply libr/core/canal.c:826]: yes. its O(n) but we should have O(1) apis in rbin at some point imho

File: libr/core/canal.c (hunk 2 above)
Comment (pelijah) [reply libr/core/canal.c:832]: `entry.*` flags are stored in the `symbols` flag space too :\

Comment (thestr4ng3r) [reply libr/core/canal.c:832]: Hmm ok yeah that sucks. Then it's ok for now I guess.

File: libr/core/canal.c [hunk 4]
```diff
@@ -819,8 +823,21 @@ static int __core_anal_fcn(RCore *core, ut64 at, ut64 from, int reftype, int dep
...
+					r_list_foreach (syms, iter, sym) {
+						if (sym->vaddr == fcn->addr) {
+							free (new_name);
+							new_name = r_str_newf ("sym.%s", sym->name);
+							break;
+						}
+					}
+				}
+				free (fcn->name);
+				fcn->name = new_name;
```
Comment (pelijah) [inline libr/core/canal.c:840]: @radare Not sure that explict overriding of a name field is a good practice

File: libr/core/canal.c [hunk 5]
```diff
@@ -819,8 +823,21 @@ static int __core_anal_fcn(RCore *core, ut64 at, ut64 from, int reftype, int dep
...
 			if (f && f->name && strncmp (f->name, "sect", 4)) { /* Check if it's already flagged */
-				R_FREE (fcn->name);
-				fcn->name = strdup (f->name);
+				const char *new_name = strdup (f->name);
+				if (is_entry_flag (f)) {
+					RListIter *iter;
+					RBinSymbol *sym;
+					const RList *syms = r_bin_get_symbols (core->bin);
+					r_list_foreach (syms, iter, sym) {
+						if (sym->vaddr == fcn->addr && !strcmp (sym->type, R_BIN_TYPE_FUNC_STR)) {
```
Comment (pelijah) [inline libr/core/canal.c:832]: Got it. This check doesn't work if base address is set.

**Comment by pelijah** [comment]
Can't reproduce that `[XX] db/archos/linux-x64/dbg_oo ood after open in debug mode` on my machine

**Comment by radare** [comment]
thanks! yeah i cant repro this issue either maybe random

**Pull request merged**